connector/src/main/scala/com/microsoft/kusto/spark/datasource/KustoRelation.scala

```diff
@@ -38,79 +38,77 @@ private[kusto] case class KustoRelation(kustoCoordinates: KustoCoordinates,
-    var timedOutCounting = false
-    val forceSingleMode = readOptions.readMode.isDefined && readOptions.readMode.get == ReadMode.ForceSingleMode
-    var useSingleMode = forceSingleMode
-    var res: Option[RDD[Row]] = None
-    if (readOptions.readMode.isEmpty){
-      var count = 0
-      try {
-        count = KDSU.
-          estimateRowsCount(kustoClient.engineClient, query, kustoCoordinates.database, clientRequestProperties.orNull)
-      } catch {
-        // Assume count is high if estimation got timed out
-        case e: Exception =>
-          if (forceSingleMode) {
-            // Throw in case user forced LeanMode
-            throw e
-          }
-          // By default we fallback to distributed mode
-          timedOutCounting = true
-      }
-      if (count == 0 && !timedOutCounting) {
-        res = Some(sparkSession.emptyDataFrame.rdd)
-      } else {
-        // Use distributed mode if count is high or in case of a time out
-        useSingleMode =  !(timedOutCounting || count > KustoConstants.DirectQueryUpperBoundRows)
-      }
+    val kustoClient = KustoClientCache.getClient(kustoCoordinates.clusterUrl, authentication,
+      kustoCoordinates.ingestionUrl, kustoCoordinates.clusterAlias)
+    val isUserOptionForceSingleMode = readOptions.readMode.contains(ReadMode.ForceSingleMode)
+    val storageParameters = maybeStorageParameters.getOrElse(kustoClient.getTempBlobsForExport)
+    val (useSingleMode, estimatedRecordCount) = readOptions.readMode match {
+      // if the user provides a specific option , this is to be used no matter what
+      case Some(_) => (isUserOptionForceSingleMode, -1)
+      // If there is no option mentioned , then estimate which option to use
+      // Count records and see if we wat a distributed or single mode
+      case None => val estimatedRecordCountResult =  Try(KDSU.
+        estimateRowsCount(kustoClient.engineClient, query, kustoCoordinates.database, clientRequestProperties.orNull))
+        estimatedRecordCountResult match {
+          // if the count is lss than the 5k threshold,use single mode.
+          case Success(recordCount) => (recordCount <= KustoConstants.DirectQueryUpperBoundRows, recordCount)
+          // A case of query timing out. ForceDistributedMode will be used here
+          case Failure(_) => (false, -1)
+        }
     }
-
-    var exception: Option[Exception] = None
-    if(res.isEmpty) {
-      if (useSingleMode) {
-        try {
-          res = Some(KustoReader.singleBuildScan(
-            kustoClient,
-            KustoReadRequest(sparkSession, cachedSchema, kustoCoordinates, query, authentication, timeout, clientRequestProperties, requestId),
-            KustoFiltering(requiredColumns, filters)))
-        } catch {
-          case ex: Exception => exception = Some(ex)
+    // if the selection yields 0 rows there are no records to scan and process
+    if(estimatedRecordCount == 0){
```

Comment:
This alarmed me for a second as i forgot my logic myself ...

Please comment it that this is zero only if a real count() kql was issue, if estimation returns zero or null estimatedRecordCount will equal 1